src/classes/BDI_GAUAllocationsUtil.cls

```diff
@@ -0,0 +1,252 @@
+     */
+    public static Map<Id, DataImport__c> validateDIGAUPercentages(BDI_DataImportService bdi,
+                                                                    DataImport__c[] dataImports) {
+        Map<Id, DataImport__c> diWithValidGAUToProcess = new Map<Id, DataImport__c>();
+
+        // Loop through the DIs to determine if they exceed the allocation percentage max.
+        for (DataImport__c di : dataImports) {
+            Decimal totalPercent = 0;
+
+            // Add to the total percent if the value is not null.
+            if (di.GAU_Allocation_1_Percent__c != null) {
+                totalPercent += di.GAU_Allocation_1_Percent__c;
+            }
+            if (di.GAU_Allocation_2_Percent__c != null) {
+                totalPercent += di.GAU_Allocation_2_Percent__c;
+            }
+
+            // if the total percent is over 100 then fail the record and log an error.
+            if (totalPercent > 100) {
+                bdi.LogBDIError(di,label.bdiErrorGAUAllocationOver100,'GAU_Allocation_1_Import_Status__c');
+            } else {
+                // If the percent is less than 100 or they were null then pass it through for
+                // more processing
+                diWithValidGAUToProcess.put(di.Id,di);
+            }
+        }
+
+        return diWithValidGAUToProcess;
+    }
+
+    /**
+     * @description Inserting the new GAU Allocations defined on the Data Import record will likely cause issues if
+     * there are any existing Allocations on the Donation listed in DonationImported__c or if there are any allocations
+     * linked to an existing payment for the Donation listed in DonationImported__c.  This method will add an error
+     * and fail the Data Import record if there are any existing Allocations located in either place.
+     * @param bdi the BDI_DataImportService instance that is being used for this context.
+     * @param dataImports {List<DataImport__c>} the data import records to search for existing Allocations for.
+     */
+    public static void checkForExistingAllocations(BDI_DataImportService bdi, DataImport__c[] dataImports) {
+
+        Map<Id,DataImport__c[]> disByOpptId = new Map<Id,DataImport__c[]>();
+
+        // Loops through data import records and if the donation imported is not null then add to list for review of
+        // existing allocations.
+        for (DataImport__c di : dataImports) {
+            if (di.DonationImported__c != null) {
+
+                if (disByOpptId.get(di.DonationImported__c) != null) {
+                    disByOpptId.get(di.DonationImported__c).add(di);
+                } else {
+                    disByOpptId.put(di.DonationImported__c,new DataImport__c[]{di});
+                }
+            }
+        }
+
+        // If there are disByOpptId then query existing Allocations
+        if (!disByOpptId.isEmpty()) {
+            Set<Id> opptIds = disByOpptId.keySet();
```

Comment:
checkmarks usually requires to check for != null every set/list that is part of where clause.

Reply:
@grivero Based on checking that the disByOpptId isn't empty then the keyset would have to be not null as well.  Lets see if it complains.